Import refund data from Shopify to Solidus

thisiscab opened the pull request:
This will let us display refunded products from Shopify to Solidus

https://app.clubhouse.io/dynamomtl/story/24/as-a-csr-i-am-able-to-refund-a-purchase-made-in-stores-via-solidus-so-that-i-don-t-need-to-use-shopify

thisiscab committed:
Import refund data from Shopify to Solidus

This will let us display refunded products from Shopify to Solidus

https://app.clubhouse.io/dynamomtl/story/24/as-a-csr-i-am-able-to-refund-a-purchase-made-in-stores-via-solidus-so-that-i-don-t-need-to-use-shopify

thisiscab committed:
Merge branch 'feature/shopify/order-sync-from-hook' into feature/shopify/import-refunds

thisiscab committed:
Add the routes to allow import refunds for hook

thisiscab committed:
Fix specs that were broken due to changes

thisiscab committed:
Fix rubocop complains

thisiscab committed:
Merge branch 'master' into feature/shopify/import-refunds

thisiscab committed:
Add class that will import refund from Shopify orders

thisiscab committed:
Add logic to not re-send refund to Shopify

The problem I am facing is that we wanted to refund orders on Shopify
when they were refunded on Solidus. Which works fine at the moment.

But now that we want it the other way around, which is saving the refund
on Solidus when it's done in Solidus, this gets a little bit more crazy.

We will receive the event and we will proceed to do the Solidus refund
when a Shopify refund is done. When creating a Solidus refund at a later
step it tries to do a refund on the Gateway, but in our scenario the
Gateway is Shopify.

Which makes the following scenario happen: Shopify Refund -> Solidus
Refund -> ShopifyRefund and then it crashes because there are nothing
more to refund on Shopify.

This is a WIP solution for this problem.

thisiscab committed:
Try to set the order to returned

Since we are skipping a few state for the customer_return and
return_authorization, we have to go back the tree and try to refund all
the inventory_units of a refund and then try to set the order state to
return.

thisiscab commented:
Future me, please right more specs and make the current spec run.
K thanks!

thisiscab commented:
@braidn @DanielWright 

My future self will write more tests but this is still QAble, so if you're interested please be my guess. Thanks! :D

thisiscab committed:
Remove duplicated shared context

DanielWright commented on app/controllers/spree/retail/shopify/hooks/refunds_controller.rb:
```diff
@@ -0,0 +1,24 @@
+          end
+
+          def shopify_order_id
+            params[:order_id]
```
Is it possible for this to not be set, or to be otherwise invalid? Does the Shopify hook care what the response is (500 vs 404 vs 422 vs etc.)?

thisiscab replied:
If the response code is not 200 then Shopify re-sends the event.

I don't see situations here the `order_id` params wouldn't be set on the Shopify side of it. What would you suggest here?

DanielWright replied:
Yeah, it seems impossible, since the route wouldn't match without a legit order ID. I guess I might be concerned the Shopify order couldn't be found?

In general I'm trying to ensure you've thought through the possible error cases and how this action ought respond in those cases.

DanielWright commented on app/controllers/spree/retail/shopify/hooks/refunds_controller.rb:
```diff
@@ -0,0 +1,24 @@
+        class RefundsController < HooksController
+          def create
+            shopify_refund = ShopifyAPI::Refund.find(shopify_refund_id, params: { order_id: shopify_order_id })
+            Spree::Retail::Shopify::GenerateRefundOrder.new(shopify_refund).process
```
Can this fail?

thisiscab replied:
Yes this could fail, I could surround it by a catch block and output the errors in a log?

DanielWright replied:
What do you think would be the most useful way, or the most intuitive way, for this action to respond? Should it return 200 no matter what so Shopify doesn't call the webhook again?

thisiscab replied:
I think that Shopify tries 3 times to receive a 200 response after that it stops.

I feel like we should log the problem in a log file and let it be if it fails more than 3 times. I don't know what scenario would cause that problem, so by logging it we would be able to implement a better behavior?

DanielWright replied:
Based on my reading of `GenerateRefundOrder`, there are lots of reasons it _might_ fail. But without specific insight into how we might rationally recover, I think your idea is a good one. Assuming the webhook method of synching a refund fails, can it be imported separately?

thisiscab replied:
Right now, there is no easy way to import it separately. 

We could have a task that loops over all the ShopifyOrder and check if there is a refund order associated with it, if there is none then we try to import it but I foresee a lot of problem with that idea. 

I am unsure how much time we want to spend on that.

DanielWright replied:
Let's wait until it's motivated by pain. I think at some point we'll need to attack the idea of importing a refund into Solidus, but that day is not today!

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -1,6 +1,31 @@
+    delegate :pos_refunded, to: :reimbursement
+
+    def perform!
+      if pos_refunded
```
A little bit of Rails' syntactic sugar is defining a predicate method for boolean model attributes. So, for the `pos_refunded` attribute, Rails defines the method `pos_refunded?`. What do you think of the convention of using the predicate form, rather than inferring it from the accessor?

thisiscab replied:
That's what I usually do, but I had a problem that I posted in the dev channel where

`pos_refunded` was returning true/false and `pos_refunded?` was returning false when the flag what set to true.

This made no sense and I received no feedback so I proceeded with the way that I know what working for sure.

I'll spend more time and try to figure it out.

DanielWright replied:
I notice that your migration uses the column type `:bool`, which I didn't know was a valid column type. Maybe Rails only picks up on `:boolean`?

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -1,6 +1,31 @@
+
+        return true
+      else
+        super
```
Could this nesting be eliminated by converting this to a guard clause? For example: `return super unless pos_refunded?`

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -1,6 +1,31 @@
+
+    def return_all_inventory_unit!(reimbursement)
+      reimbursement.return_items.each do |return_item|
+        return_item.inventory_unit.return!
```
What happens if any of the IUs fail the state transition?

thisiscab replied:
Then it goes boom.

DanielWright replied:
Is that in line with Solidus's typical behaviour in cases like these?

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -1,6 +1,31 @@
+    end
+
+    def try_set_order_to_return!(reimbursement)
+      reimbursement.order.return!
```
Does the Reimbursement model not provide an interface for this? Could we define one, to avoid Demeter sad-making?

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -1,6 +1,31 @@
+    end
+
+    def try_set_order_to_return!(reimbursement)
+      reimbursement.order.return!
```
Also, what happens if this fails? You've used the whiny trigger, but nothing seems to rescue the exception.

thisiscab replied:
if it fails, it goes boom.

Suggestions?

I may be repeating myself here, but I just want to be clear that Shopify sends us a specific amount of data and we are managing only a specific case of what could be happening. They are planning to instruct the POS user to not do things that we are not supporting.

In the scenario that it fails, this will be re-tried 3 time where we will be logging those error. So we can eventually fix those scenarios if they ever happen.

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -0,0 +1,86 @@
+        end
+
+        def process
+          order = find_order(shopify_refund)
```
I find local assignments to be a faint code smell. They introduce just a little bit of extra complexity to a method, and a little bit more statefulness to the object. Sometimes local assignments can't be helped, but this example could be refactored as a private `order` method. Same thing with `return_items` below. You may find if the complexity of this method, and the class overall, is reduced dramatically as you do things like this.

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -0,0 +1,86 @@
+        private
+
+        def find_inventory_by_shopify_variant_id(order, shopify_variant_id)
+          order.shipments.first.inventory_units.find { |unit| unit.variant.pos_variant_id.to_i == shopify_variant_id.to_i }
```
There's a lot of reaching through objects here. Anything to be done about that?

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -0,0 +1,86 @@
+        end
+
+        def stock_location_to_refund
+          Spree::StockLocation.first
```
Why the first stock location? Should this be more clever about selecting the "right" stock location?

thisiscab replied:
What is the "right" stock location? I feel like it depends on the host implementation and that method should be decorated? 

Would you prefer having it configured in the initializer?

DanielWright replied:
It feels kinda wrong to expect the host app to decorate the class to get the correct behaviour. Do you think this is okay for a first iteration, and we can add configurability as a separate future concern?

(Separate musing: do you think we ought to have a legit configuration panel in the backoffice?)

thisiscab replied:
If the time concern and the fact that we mainly want to have something that work for V1 around then I would say it's good enough for the first Iteration.

If we say that we want to do it properly and we want to really make it open-sourced, then I would say that we really really need a configuration panel and that there is quite a bit of stuff that needs be improved in order to make it as a proper gem.

It does feel wrong to rely on the host app to decorate the app, but we've been doing that since the beginning of this gem, so I'm unsure what to say about this.

DanielWright replied:
Cool, we're on the same page. Shall I add something to Clubhouse, or here as an issue?

thisiscab replied:
I shall recommend you create an issue :)

DanielWright replied:
👍

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -0,0 +1,86 @@
+        end
+
+        def reimbursement_type
+          Spree::ReimbursementType.first
```
Same question as above.

thisiscab replied:
Same response as above

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -0,0 +1,86 @@
+        end
+
+        def return_reason
+          Spree::ReturnReason.first
```
Same question as below.

thisiscab replied:
Same response as below

DanielWright replied:
😆

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -0,0 +1,86 @@
+            return_items << create_return_item(inventory_unit)
+          end
+
+          create_return_authorization(order, return_items)
```
I'm concerned about the failure cases here. Many of these methods use the non-whiny commands (`create` vs `create!`), which could "silently" fail to persist. How are failures handled?

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -0,0 +1,86 @@
+            inventory_unit: inventory_unit,
+            preferred_reimbursement_type: reimbursement_type
+          )
+          return_item.accept!
```
This is a nice use-case for [`Object#tap`](http://ruby-doc.org/core-2.3.1/Object.html#method-i-tap). You could express these two lines more expressively with `return_item.tap(&:accept!)`.

DanielWright commented on spec/models/spree/retail/shopify/generate_refund_order_spec.rb:
```diff
@@ -0,0 +1,39 @@
+Spree.describe Spree::Retail::Shopify::GenerateRefundOrder, type: :model do
+  include_context 'shopify_request'
+
+  let!(:store) { create :store }
```
Yikes! That is a _lot_ of dependencies!

thisiscab replied:
How would you fix that? Separating the class into multiple smaller classes?

DanielWright replied:
I probably would, yeah. To have full test coverage would require a _ton_ of specs to cover the myriad control paths, and there's a lot of database access so they'd be slow.

But I recognize the magnitude of effort involved in such a refactoring. Do you feel comfortable with the object as-is? Do you think the effort can be safely deferred?

thisiscab replied:
I don't feel confortable in refactoring it, since it will require a lot of work.

We would also need to refactor how the similar order class work, because why do one and not do the other?

I feel comfortable with the object as it is right now because it works with real data but I do not feel comfortable about how the specs are written because, myself, I have hard-time understanding what's going on.

It's not something that can be done with a one-man effort.

DanielWright replied:
Oh, I think the order-generator should be refactored as well, although in a different way. It's out-of-scope for this PR, though. I'll add an issue for this, and for the order-generator. Cool?

DanielWright reviewed (changes requested):
Hey @cabouffard,  I've given this sucka a solid pass. I'll be around all afternoon to respond to any replies/comments you have. :boom:

thisiscab committed:
Fix the refund specs

Multiple specs were broken due to the recent refund change.

There were multiple problems.
1) The set-up of the shipping method was done before the change of
address->deliver instead of after.
2) The amounts of the order support file were different than the amount
of the refund support file. Since we are using the order support file to
create a real order, those data needed to match

thisiscab committed:
Change refunded flag to boolean

A mistake was made here and it was were setting up a wrong field.
Sometimes it would create it as a string and sometimes as a boolean.
When it was creating the field as a string, then the methode `refunded?`
was returning wrong values.

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -1,6 +1,31 @@
     delegate :pos_order_id, to: :payment
+    delegate :pos_refunded, to: :reimbursement
+
+    def perform!
```
Sorry, fresh comment here. Are you overriding an existing `perform!` method here, or is this a new interface method?

thisiscab committed:
Use guard clause instead of if else statement

DanielWright commented on app/controllers/spree/retail/shopify/hooks/refunds_controller.rb:
```diff
@@ -0,0 +1,24 @@
+    module Shopify
+      module Hooks
+        class RefundsController < HooksController
+          def create
```
Is this action idempotent?

thisiscab replied:
In the Shopify UI, you can't re-refund something that has already been re-fund. 

I am unsure of the case where we would receive twice the same event and thus having to manage the case where it has to be idempotent. But I could take a look and make it work if you think it's a good idea.

DanielWright replied:
It's more of a Murphy's Law kind of thing. Maybe the load-balancer accidentally sends the request to two web servers, and the webhook action ends up getting invoked twice, completely by accident.

I might be thinking _too_ defensively here, though. What do you think?

thisiscab replied:
I obviously initially thought that it was not necessary thus why I didn't implement it, but I see your point and I'll change it to make it idempotent.

thisiscab replied:
Fixed!

thisiscab committed:
Fix specs

thisiscab committed:
Call the method set to order to return from the reimbursement

This way Demeter won't be mad at us.

thisiscab committed:
Reduce local assignments for the refunder class

As suggested in the CR

https://github.com/glossier/solidus_retail/pull/49#discussion_r83675159

thisiscab committed:
Use tap instead of local assignments

thisiscab committed:
Make the GenerateRefundOrder idempotent + logging

thisiscab committed:
Add specs to ensure correct state change after refund

thisiscab committed:
Remove rubocop too many lines offense for gemspec file

DanielWright commented on app/models/spree/retail/shopify/generate_pos_order.rb:
```diff
@@ -109,14 +109,12 @@ def transition_order_from_cart_to_address!(order)
 
         def transition_order_from_delivery_to_payment!(order)
+          order.shipments.first.stock_location = Spree::StockLocation.find_by(admin_name: 'POPUP')
+          order.shipments.first.add_shipping_method(Spree::ShippingMethod.find_by(admin_name: 'POS'), true)
```
Will it be impossible for the order's shipments to have a shipping rate _other_ than the POS's? I assume yes, just double-checking.

thisiscab replied:
It's actually impossible for now. I feel like this way of doing it doesn't allow for much extension, but it works... :)

DanielWright commented on spec/models/spree/retail/shopify/generate_refund_order_spec.rb:
```diff
@@ -8,19 +8,26 @@
-    allow_any_instance_of(Spree::Reimbursement).to receive(:perform!).and_return(true)
+  let!(:variant) { create :variant, price: 12.00, pos_variant_id: '29637399811' }
+
+  let!(:order_response_mock) { mock_request('orders', 'orders/450789469', 'json') }
```
The number `450789469` appears a whole bunch. Possible to DRY them up?

DanielWright commented on spec/requests/shopify/export_refund_spec.rb:
```diff
@@ -14,6 +13,7 @@
              amount: refund_amount,
              payment_method: shopify_payment_method)
     end
+    let(:pos_order) { create_fulfilled_paid_shopify_order }
```
Did moving this code down the page change things? Because that seems _crazy_!

thisiscab replied:
It didn't! I don't have any reasons of why I moved it. I just did it without really thinking.

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -4,15 +4,13 @@ module RefundDecorator
+      try_set_order_to_return!(reimbursement)
+      update_order
+
+      true
```
Is this necessary to avoid returning what `update_order` returns?

DanielWright commented on app/models/spree/refund_decorator.rb:
```diff
@@ -1,10 +1,10 @@
 
     def perform!
-      super unless pos_refunded?
+      return super if reimbursement.nil? || !pos_refunded?
```
Do you think it'd be an improvement to use [the `allow_nil` option](http://api.rubyonrails.org/classes/Module.html#method-i-delegate) to the `delegate` above, in order to avoid this nil check?

DanielWright commented on app/models/spree/reimbursement_decorator.rb:
```diff
@@ -3,6 +3,10 @@ module ReimbursementDecorator
       update_attribute(:pos_refunded, true)
     end
+
+    def try_set_order_to_return!
```
Nitpicking a bit: I'm not a big fan of the `try_` prefix here, for the same reason I don't like the `Object#try` method: you're not _trying_ to trigger the return event, you _are_, and "try" evokes the "try/catch/finally" structure, and you're not rescuing the exception.

Long-winded way to say, I'd name this simply `return_order!` 😄

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -4,24 +4,31 @@ module Shopify
+        attr_reader :shopify_refund, :order
+
+        def return_items
+          @return_items ||= begin
```
Yummy! Love it. Remind me to show you `inject` sometime to make examples like this even cleaner.

thisiscab replied:
Now would be the good time, no?

DanielWright replied:
Actually, looking at it again, this isn't the _greatest_ case for inject/reduce. It's more of a map/collect:

``` ruby
return_items = shopify_refund.refund_line_items.map do |rli|
  inventory_unit = find_inventory_by_shopify_variant_id(order, rli.line_item.variant_id)
  create_return_item(inventory_unit)
end
```

This eliminates the need for declaring your own aggregator array, and of shovelling the return items onto it.

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -34,12 +34,10 @@ def find_inventory_by_shopify_variant_id(order, shopify_variant_id)
 
         def create_return_item(inventory_unit)
-          return_item = Spree::ReturnItem.create(
+          Spree::ReturnItem.create(
```
❤️

DanielWright commented on app/models/spree/retail/shopify/generate_refund_order.rb:
```diff
@@ -8,9 +8,20 @@ def initialize(shopify_refund)
+
+          return true
+
+        rescue => e
```
Any concern about indiscriminately rescuing all exceptions, vs. selectively rescuing exceptions you anticipate?

thisiscab replied:
Let's catch all the exceptions for now :)

DanielWright commented on spec/models/spree/retail/shopify/generate_refund_order_spec.rb:
```diff
@@ -43,6 +43,11 @@
 
+    it 'sets the order state to returned' do
+      subject
+      expect(last_order).to be_returned
```
Nice!

DanielWright commented on solidus_retail.gemspec:
```diff
@@ -2,6 +2,7 @@
 $:.push File.expand_path('../lib', __FILE__)
 require 'solidus_retail/version'
 
+# rubocop:disable BlockLength
```
Rubocop, you so crazy.

DanielWright commented on spec/models/spree/retail/shopify/generate_pos_order_spec.rb:
```diff
@@ -41,7 +41,7 @@
     it 'ships the shipments' do
       subject
-      expect(last_order.shipments.state).to eql('shipped')
+      expect(last_order.shipments.first.state).to eql('shipped')
```
Does the state-machine make a `shipped?` predicate available on the shipment? Could this be more clearly expressed as `expect(last_order.shipments).to all(be_shipped)`?

thisiscab replied:
TIL about `all`

DanielWright reviewed (changes requested):
I know, I know. You mustn't be thrilled at 12 more comments. Sorry! Nothing major here, I promise. Mostly areas that could use a little readability/maintainability care, and some questions. Looks really solid!

thisiscab committed:
Merge branch 'master' into feature/shopify/import-refunds

thisiscab committed:
DRY up the shopify order id

thisiscab committed:
Allow the delegate to return nil instead of throwing error

thisiscab committed:
Change the return order method name

Here is DW explanation

Nitpicking a bit: I'm not a big fan of the try_ prefix here, for the
same reason I don't like the Object#try method: you're not trying to
trigger the return event, you are, and "try" evokes the
"try/catch/finally" structure, and you're not rescuing the exception.

Which I agree with

thisiscab committed:
Change a spec to be more readable

thisiscab commented:
Changed code to reflect most of your suggestions, I think we're good for a 👍 here?

DanielWright commented:
:shipit:
